commit 8a4e19abaecbe314e37fcdbfd1dfff3c47457223 (refs/changes/51/551/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-09-29T16:13:58-07:00 (3 years ago)
    
    joyent/node-cueball#30 hang when max == count and everything is dead

diff --git a/lib/utils.js b/lib/utils.js
index 17588ba..6a22339 100644
--- a/lib/utils.js
+++ b/lib/utils.js
@@ -194,13 +194,11 @@ function planRebalance(inSpares, dead, target, max, singleton) {
 		var count = done + replacements - i;
 		var empties = keys.filter(function (kk) {
 			if (singleton) {
-				return (dead[kk] !== true && (
-				    wantedSpares[kk] === undefined ||
-				    wantedSpares[kk] === 0));
+				return (dead[kk] !== true &&
+				    wantedSpares[kk] === undefined);
 			} else {
 				return (dead[kk] !== true ||
-				    wantedSpares[kk] === undefined ||
-				    wantedSpares[kk] === 0);
+				    wantedSpares[kk] === undefined);
 			}
 		});
 		if (count + 1 <= max) {
diff --git a/test/utils.test.js b/test/utils.test.js
index 16950f1..b2d09ab 100644
--- a/test/utils.test.js
+++ b/test/utils.test.js
@@ -263,3 +263,22 @@ mod_tape.test('rebalance: dead, backend starvation', function (t) {
 	t.deepEqual(plan.add, ['b2', 'b2', 'b2']);
 	t.end();
 });
+
+mod_tape.test('bug #30', function (t) {
+	var spares = {
+		'16uN6JsJFild9cHyl2+LSyRHmNc=': ['c1'],
+		'c7QG0UOYCpm6m/hYUX0jBenbM70=': ['c2'],
+		'ashWtupYHh1QH33UP/T2+6hvi8c=': [],
+		'4QMg6SChOmtF8s6lfK32lLoKUFs=': []
+	};
+	var dead = {
+		'c7QG0UOYCpm6m/hYUX0jBenbM70=': true,
+		'16uN6JsJFild9cHyl2+LSyRHmNc=': true,
+		'4QMg6SChOmtF8s6lfK32lLoKUFs=': true,
+		'ashWtupYHh1QH33UP/T2+6hvi8c=': true
+	};
+	var plan = mod_utils.planRebalance(spares, dead, 3, 4);
+	t.deepEqual(plan.remove, []);
+	t.deepEqual(plan.add, ['ashWtupYHh1QH33UP/T2+6hvi8c=']);
+	t.end();
+});
